### PR TITLE
Add a RISC-V standard performance monitoring interface to Toooba, and hookup lot of events to it.

### PR DESCRIPTION
**Pull request by jonwoodruff**
This has been tested under FreeBSD.  Our fork of BBL maps a static 29 events and enables the counters for userspace access.  Executables then access them using the unprivileged instructions directly.  As far as we know, all standard RISC-V operations and protections for these registers are supported.

### CODE REVIEW

File: src_Core/RISCY_OOO/procs/RV64G_OOO/CommitStage.bsv [hunk 1]
```diff
@@ -1037,6 +1087,23 @@ module mkCommitStage#(CommitInput inIfc)(CommitStage);
                 supComUserCnt.incr(1);
             end
         end
+`endif
+`ifdef PERFORMANCE_MONITORING
+        EventsCore events = unpack(0);
+        events.evt_BRANCH = brCnt;
+        events.evt_JAL = jmpCnt;
+        events.evt_JALR = jrCnt;
+        events.evt_AUIPC = auipcCnt; // XXX
```
Comment (jrtc27) [inline src_Core/RISCY_OOO/procs/RV64G_OOO/CommitStage.bsv:1096]: Is this just XXX for downstream where we conflate AUIPC and AUIPCC?

File: src_Core/RISCY_OOO/procs/RV64G_OOO/MemExePipeline.bsv [hunk 2]
```diff
@@ -307,18 +326,24 @@ module mkMemExePipeline#(MemExeInput inIfc)(MemExePipeline);
...
+            if (waitSt.shiftedBE == -1) events.evt_MEM_CAP_STORE = 1;
+            events.evt_STORE_WAIT = saturating_truncate(lat);
+            events_reg[2] <= events;
 `endif
             // now figure out the data to be written
             Vector#(LineSzData, ByteEn) be = replicate(replicate(False));
             Line data = replicate(0);
             be[waitSt.offset] = waitSt.shiftedBE;
-            data[waitSt.offset] = waitSt.shiftedData;
+            data[waitSt.offset] = waitSt.shiftedData; //XXX I guess this doesn't work with capabilities?  Maybe we don't build TSO?
```
Comment (jrtc27) [inline src_Core/RISCY_OOO/procs/RV64G_OOO/MemExePipeline.bsv:346]: CHERI

File: src_Core/RISCY_OOO/procs/lib/ProcTypes.bsv [hunk 3]
```diff
@@ -982,3 +1075,76 @@ function Fmt showInst(Instruction inst);
...
+   SupCnt evt_D_BUSY_NO_CONSUME; // XXX
+   SupCnt evt_1_BUSY_NO_CONSUME; // XXX
+   SupCnt evt_2_BUSY_NO_CONSUME; // XXX
+   SupCnt evt_3_BUSY_NO_CONSUME; // XXX
+   SupCnt evt_IMPRECISE_SETBOUND; // XXX
+   SupCnt evt_UNREPRESENTABLE_CAP; // XXX
+   SupCnt evt_MEM_CAP_LOAD;
+   SupCnt evt_MEM_CAP_STORE;
+   SupCnt evt_MEM_CAP_LOAD_TAG_SET;
+   SupCnt evt_MEM_CAP_STORE_TAG_SET;
```
Comment (jrtc27) [inline src_Core/RISCY_OOO/procs/lib/ProcTypes.bsv:1113]: None of these should exist, especially the CHERI ones upstream

File: src_Core/RISCY_OOO/procs/lib/LLCache.bsv [hunk 4]
```diff
@@ -1,17 +1,30 @@
 
 // Copyright (c) 2017 Massachusetts Institute of Technology
-// 
+//
+//-
+// RVFI_DII + CHERI modifications:
```
Comment (jrtc27) [inline src_Core/RISCY_OOO/procs/lib/LLCache.bsv:5]: No RVFI_DII or CHERI here

**Comment by quark17** [comment]
Am I missing something or does this not build?  This PR adds imports of `PerformanceMonitor` and `SpecialRegs` (when `PERFORMANCE_MONITORING` is defined, which it is in the Makefile) but the PR does not add those files, so I get errors:
```
Error: "../../src_Core/CPU/Core.bsv", line 48, column 8: (S0000)
  Cannot find package `PerformanceMonitor'
Error: "../../src_Core/CPU/Core.bsv", line 49, column 8: (S0000)
  Cannot find package `SpecialRegs'
...
```

Also, were @jrtc27 's comments addressed?  Was this perhaps not ready to be merged?

**Comment by jonwoodruff** [comment]
Hey Julie!
There is now a subrepo here, so "git submodule update --init --recursive" is now required.
Apologies for sneaking this in! Is there some way to make submodules more "invisible" to the user?

The issues pointed out by @jrtc27 were resolved in the next pull request, as this one was merged post-haste.

**Comment by quark17** [comment]
Ah, I see.  Sorry!

I believe that BSC did report a warning that a search directory that didn't exist.  I should have paid more attention.  Perhaps this warning number could be raised to an error (via command-line flags in the build Makefiles), but maybe better would be a target in the Makefiles that looks for the directories and report an error (if missing) that includes the `git submodule` commands to initialize the submodules.  It could also be mentioned in the README.

**Comment by jonwoodruff** [comment]
Ok, I've made a new pull request with the Makefile trick as well as a small update to the readme.

**Pull request merged**